ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
+	for (i = 0; i < keys_len; i++) {
+		VALUE rb_key = rb_ary_entry(rb_keys, i);
+		Check_Type(rb_key, T_STRING);
 
-  while ((result = memcached_fetch_result(ctx->memc, NULL, &rc))) {
-    handle_memcached_return(rc);
+		c_keys[i] = RSTRING_PTR(rb_key);
+		c_lengths[i] = RSTRING_LEN(rb_key);
+	}
 
-    VALUE rb_key = rb_str_new(memcached_result_key_value(result), memcached_result_key_length(result));
-    VALUE rb_value = rb_str_new(memcached_result_value(result), memcached_result_length(result));
+	rc = memcached_mget(mc, c_keys, c_lengths, (size_t)keys_len);
+	rb_memcached_error_check(rc);
 
-    rb_hash_aset(rb_values, rb_key, rb_value);
-    memcached_result_free(result);
-  }
+	rb_result = rb_hash_new();
+	rc = memcached_fetch_execute(mc, callbacks, (void *)rb_result, 1);
+	rb_memcached_error_check(rc);
 
-  return rb_values;
+	return rb_result;
 }
 
 static VALUE
-rb_connection_delete(VALUE self, VALUE key)
+rb_connection_delete(VALUE self, VALUE rb_key)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  char *mkey = StringValuePtr(key);
-  memcached_return_t rc = memcached_delete(ctx->memc, mkey, strlen(mkey), (time_t)0);
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+
+	rc = memcached_delete(mc, RSTRING_PTR(rb_key), RSTRING_LEN(rb_key), (time_t)0);
+	rb_memcached_return(rc);
 }
 
 static VALUE
-rb_connection_add(VALUE self, VALUE rb_key, VALUE rb_value, VALUE ttl, VALUE flags)
+rb_connection_add(VALUE self, VALUE rb_key, VALUE rb_value, VALUE rb_ttl, VALUE rb_flags)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  char *mkey = StringValuePtr(rb_key);
-  char *mvalue = StringValuePtr(rb_value);
-
-  memcached_return_t rc = memcached_add(ctx->memc, mkey, strlen(mkey), mvalue, strlen(mvalue), NUM2INT(ttl), NUM2INT(flags));
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_add(mc,
+			RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+			RSTRING_PTR(rb_value), RSTRING_LEN(rb_value),
+			FIX2INT(rb_ttl), FIX2INT(rb_flags)
+	);
+
+	rb_memcached_return(rc);
 }
 
 static VALUE
 rb_connection_inc(VALUE self, VALUE rb_key, VALUE rb_value)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  char *mkey = StringValuePtr(rb_key);
-  uint32_t offset = NUM2UINT(rb_value);
-  uint64_t new_number;
-
-  memcached_return_t rc = memcached_increment(ctx->memc, mkey, strlen(mkey), offset, &new_number);
-  handle_memcached_return(rc);
-  return INT2NUM(new_number);
+	memcached_st *mc;
+	memcached_return_t rc;
+	uint64_t result;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_FIXNUM);
+
+	rc = memcached_increment(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		(uint32_t)FIX2ULONG(rb_value), &result
+	);
+	rb_memcached_error_check(rc);
+	return ULL2NUM(result);
```

Comment:
`ULL2NUM` will properly wrap a *large* 64-bit return value. We could have truncated them previously.